### PR TITLE
op-deployer: Support overriding proof parameters

### LINKED ISSUE #12711: Allow fault proof parameters to be configured as overrides
Currently, all the fault proof parameters are hardcoded. This includes:

- faultGameMaxClockDuration
- faultGameWithdrawalDelay
- disputeGameFinalityDelaySeconds
- faultGameClockExtension
- proofMaturityDelaySeconds

Customers are asking to override these so that they can deploy testnets with shorter withdrawal windows. To fix this, we should read these values from the override config rather than hardcoding them.

### PR DESCRIPTION
**Pull request by mslipper**
This PR adds support for overiding proofs params using the `GlobalDeployOverrides` and `DeployOverrides` fields. Previously, these values were hardcoded. To prevent a recursive import, I refactored the `standard.go` file into a standalone package.

Closes https://github.com/ethereum-optimism/optimism/issues/12711.

### CODE REVIEW

File: op-deployer/pkg/deployer/bootstrap/opcm.go [hunk 1]
```diff
@@ -28,10 +30,16 @@ import (
...
+	L1RPCUrl                        string
+	PrivateKey                      string
+	Logger                          log.Logger
+	ArtifactsLocator                *opcm.ArtifactsLocator
+	WithdrawalDelaySeconds          uint64
+	MinProposalSizeBytes            uint64
+	ChallengePeriodSeconds          uint64
+	ProofMaturityDelaySeconds       uint64
+	DisputeGameFinalityDelaySeconds uint64
+	MIPSVersion                     uint64
```
Comment (bitwiseguy) [inline op-deployer/pkg/deployer/bootstrap/opcm.go]: Instead of listing these fields out individually, could we nest the `SuperchainProofParams` struct ([link](https://github.com/ethereum-optimism/optimism/blob/feat/override-proof-params/op-deployer/pkg/deployer/pipeline/implementations.go#L14-L21))? Would help reduce redundant code

File: op-deployer/pkg/deployer/pipeline/opchain.go [hunk 2]
```diff
@@ -171,27 +180,54 @@ func DeployOPChainGenesisStrategy(env *Env, intent *state.Intent, st *state.Stat
...
+
+	return opcm.DeployOPChainInput{
+		OpChainProxyAdminOwner:       thisIntent.Roles.L1ProxyAdminOwner,
+		SystemConfigOwner:            thisIntent.Roles.SystemConfigOwner,
+		Batcher:                      thisIntent.Roles.Batcher,
+		UnsafeBlockSigner:            thisIntent.Roles.UnsafeBlockSigner,
+		Proposer:                     thisIntent.Roles.Proposer,
+		Challenger:                   thisIntent.Roles.Challenger,
+		BasefeeScalar:                1368,
+		BlobBaseFeeScalar:            801949,
```
Comment (bitwiseguy) [inline op-deployer/pkg/deployer/pipeline/opchain.go]: Where do these values come from? Could they be moved into a standards or defaults file?

File: op-deployer/pkg/deployer/pipeline/opchain.go [hunk 3]
```diff
@@ -171,27 +180,54 @@ func DeployOPChainGenesisStrategy(env *Env, intent *state.Intent, st *state.Stat
...
+		L2ChainId:                    chainID.Big(),
+		OpcmProxy:                    st.ImplementationsDeployment.OpcmProxyAddress,
+		SaltMixer:                    st.Create2Salt.String(), // passing through salt generated at state initialization
+		GasLimit:                     60_000_000,
```
Comment (bitwiseguy) [inline op-deployer/pkg/deployer/pipeline/opchain.go]: Same as previous comment: wondering if this could be moved into a standards/defaults file?

Will also note that on op-mainnet we set the SystemConfig.gasLimit to 60M, but that was a hack to set the gas target to 10M, then we override that limit by setting the `GETH_MINER_EFFECTIVEGASLIMIT = 30_000_000` on all sequencers. See details of the calculations in [this doc](https://www.notion.so/oplabs/RB-000-Increase-gas-limit-target-10cf153ee1628011b78bd6d5ba8e34d2). If we want to default to SystemConfig.gasLimit of 60M, should probably also ensure that env var override setting is the default as well. Holocene is supposed to make changes to the SystemConfig contract that will eliminate the need for this hack and the env var override

**Review by bitwiseguy — APPROVED** [review]
Go code looks good to me. Left a few minor comments

File: op-deployer/pkg/deployer/pipeline/opchain.go (hunk 3 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/pipeline/opchain.go]: Yeah good call. Let's discuss this in a thread on Discord.

File: op-deployer/pkg/deployer/pipeline/opchain.go (hunk 2 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/pipeline/opchain.go]: I can move them into the standard file.

File: op-deployer/pkg/deployer/bootstrap/opcm.go (hunk 1 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/bootstrap/opcm.go]: good idea

File: op-deployer/pkg/deployer/bootstrap/opcm.go [hunk 4]
```diff
@@ -59,6 +63,30 @@ func (c *OPCMConfig) Check() error {
...
+		c.ChallengePeriodSeconds = standard.ChallengePeriodSeconds
+	}
+
+	if c.ProofMaturityDelaySeconds == 0 {
+		c.ProofMaturityDelaySeconds = standard.ProofMaturityDelaySeconds
+	}
+
+	if c.DisputeGameFinalityDelaySeconds == 0 {
+		c.DisputeGameFinalityDelaySeconds = standard.DisputeGameFinalityDelaySeconds
+	}
```
Comment (mds1) [inline op-deployer/pkg/deployer/bootstrap/opcm.go:84]: Are there use cases for these values being 0 on devnets that are blocked by this approach?

File: op-deployer/pkg/deployer/integration_test/apply_test.go [hunk 5]
```diff
@@ -613,6 +619,126 @@ func TestApplyGenesisStrategy(t *testing.T) {
...
+
+	lgr := testlog.Logger(t, slog.LevelDebug)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	depKey := new(deployerKey)
+	l1ChainID := big.NewInt(77799777)
+	dk, err := devkeys.NewMnemonicDevKeys(devkeys.TestMnemonic)
+	require.NoError(t, err)
```
Comment (mds1) [inline op-deployer/pkg/deployer/integration_test/apply_test.go:633]: Seems all the tests in this file have a lot of boilerplate that gets duplicated, would be nice to clean that up a bit (can be a separate PR or this one)

File: op-deployer/pkg/deployer/integration_test/apply_test.go [hunk 6]
```diff
@@ -613,6 +619,126 @@ func TestApplyGenesisStrategy(t *testing.T) {
...
+		"proofMaturityDelaySeconds":               standard.ProofMaturityDelaySeconds + 1,
+		"disputeGameFinalityDelaySeconds":         standard.DisputeGameFinalityDelaySeconds + 1,
+		"mipsVersion":                             standard.MIPSVersion + 1,
+		"disputeGameType":                         standard.DisputeGameType, // This must be set to the permissioned game
+		"disputeAbsolutePrestate":                 common.Hash{'A', 'B', 'S', 'O', 'L', 'U', 'T', 'E'},
+		"disputeMaxGameDepth":                     standard.DisputeMaxGameDepth + 1,
+		"disputeSplitDepth":                       standard.DisputeSplitDepth + 1,
+		"disputeClockExtension":                   standard.DisputeClockExtension + 1,
+		"disputeMaxClockDuration":                 standard.DisputeMaxClockDuration + 1,
+		"dangerouslyAllowCustomDisputeParameters": true,
```
Comment (mds1) [inline op-deployer/pkg/deployer/integration_test/apply_test.go:659]: I like this, and wonder if we should do something similar for the L1 and L2 contract locator tags also, to discourage changing those

File: op-deployer/pkg/deployer/bootstrap/opcm.go (hunk 4 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/bootstrap/opcm.go:84]: Not as far as I know. The values can't be zero because they will fail validation in the contracts themselves.

File: op-deployer/pkg/deployer/integration_test/apply_test.go (hunk 5 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/integration_test/apply_test.go:633]: Will do as a separate PR; would like to reorder some stuff in the test file too that will make a big diff.

File: op-deployer/pkg/deployer/integration_test/apply_test.go (hunk 6 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/integration_test/apply_test.go:659]: What are you envisioning for the tags here?

Comment (mds1) [reply op-deployer/pkg/deployer/integration_test/apply_test.go:659]: Basically just want to make it very unlikely that people don't use the standard default releases (1.6.0 for L1 and 1.7.0 beta for L2). Perhaps even removing those two locator tags from the default intent to make that option less noticeable

**Review by mds1 — APPROVED** [review]

File: op-deployer/pkg/deployer/integration_test/apply_test.go (hunk 6 above)
Comment (mslipper) [reply op-deployer/pkg/deployer/integration_test/apply_test.go:659]: Ah gotcha! Yes I can modify the locators to be overrides rather than setting them raw.
